Client/Frontend/Browser/Authenticator.swift

```diff
@@ -48,6 +48,48 @@ class Authenticator {
         return self.promptForUsernamePassword(viewController, credentials: nil, protectionSpace: challenge.protectionSpace, loginsHelper: nil)
     }
 
+    static func findMatchingCredentialsForChallenge(challenge: NSURLAuthenticationChallenge, fromLoginsProvider loginsProvider: BrowserLogins) -> Deferred<NSURLCredential?> {
+        let deferred = Deferred<NSURLCredential?>()
+        loginsProvider.getLoginsForProtectionSpace(challenge.protectionSpace).upon { res in
+            var credentials: NSURLCredential? = nil
+            if let logins = res.successValue?.asArray() {
+
```

Comment:
Here — if you switch this to chaining rather than poking a `Deferred` — you'd do something like

```
guard logins.count >= 1 else {
    return deferMaybe(nil)
}
```